Rubberduck.VBEEditor/SafeComWrappers/ComWrapperEnumerator.cs

```diff
@@ -1,39 +1,143 @@
+        [PreserveSig] int Next([In] uint celt, [Out] out object rgVar, [Out] out uint pceltFetched);
+
+        [PreserveSig] int Skip([In] uint celt);
+        [PreserveSig] int Reset();
+        [PreserveSig] int Clone([Out] out IEnumVARIANT retval);
+    }
+
     public class ComWrapperEnumerator<TWrapperItem> : IEnumerator<TWrapperItem>
         where TWrapperItem : class
     {
         private readonly Func<object, TWrapperItem> _itemWrapper;
-        private readonly IEnumerator _internal;
+        private readonly IEnumVARIANT _enumeratorRCW;
+        private TWrapperItem _currentItem;
 
-        public ComWrapperEnumerator(IEnumerable source, Func<object, TWrapperItem> itemWrapper)
+        public ComWrapperEnumerator(object source, Func<object, TWrapperItem> itemWrapper)
         {
             _itemWrapper = itemWrapper;
-            _internal = source?.GetEnumerator() ?? Enumerable.Empty<TWrapperItem>().GetEnumerator();
+
+            if (source != null)
+            {
+                _enumeratorRCW = (IEnumVARIANT)IDispatchHelper.PropertyGet_NoArgs((IDispatch)source, (int)IDispatchHelper.StandardDispIds.DISPID_ENUM);
+                ((IEnumerator)this).Reset();  // precaution 
+            }
         }
 
         public void Dispose()
         {
-            // nothing to dispose here
+            if (!IsWrappingNullReference) Marshal.ReleaseComObject(_enumeratorRCW);
         }
 
-        public bool MoveNext()
+        void IEnumerator.Reset()
         {
-            return _internal.MoveNext();
+            if (!IsWrappingNullReference)
+            {
+                int hr = _enumeratorRCW.Reset();      
+                if (hr < 0)
+                {
+                    throw new ArgumentException(string.Format("Error invoking IEnumVARIANT::Reset().  COM error code is {0}", hr));
+                }
+            }
         }
+        
+        public bool IsWrappingNullReference => _enumeratorRCW == null;
+        
+        public TWrapperItem Current => _currentItem; 
+        object IEnumerator.Current => _currentItem;
 
-        public void Reset()
+        bool IEnumerator.MoveNext()
         {
-            _internal.Reset();
-        }
+            if (IsWrappingNullReference) return false;
+
+            _currentItem = null;
 
-        public TWrapperItem Current => _itemWrapper.Invoke(_internal.Current);
+            object currentItemRCW;
+            uint celtFetched;
+            int hr = _enumeratorRCW.Next(1, out currentItemRCW, out celtFetched);
+            // hr == S_FALSE (1) or S_OK (0), or <0 means error
 
-        object IEnumerator.Current => Current;
+            _currentItem = _itemWrapper.Invoke(currentItemRCW);     // creates a null wrapped reference even on end/error, just as a precaution
+            
+            if (hr < 0)
+            {
+                throw new ArgumentException(string.Format("Error invoking IEnumVARIANT::Next().  COM error code is {0}", hr));
```

Comment:
I would consider using [`Marshal. GetExceptionForHR`](https://msdn.microsoft.com/en-us/library/t8t3ykxc%28v=vs.110%29.aspx?f=255&MSPPError=-2147217396) instead here.

Reply:
sure, I'll sort that